Python-related cleanups

rossburton opened the pull request:
Ensure that the nested native CMake build uses the same Python interpreter as the outer build.

Unify python shebangs so that they use `python3` everywhere, instead of a mix of `python` and `python3`.

rossburton committed:
Unify python shebangs

As per PEP-0394[1], there is no real concensus over what binary names
Python has, specifically 'python' could be Python 3, Python 2, or not
exist.

However, everyone has a python3 interpreter and the scripts are all
written for Python 3.  Unify the shebangs so that the ~50% of shebangs
that use python now use python3.

[1] https://peps.python.org/pep-0394/

rossburton committed:
[cmake] Pass PYTHON_EXECUTABLE to native builds

Ensure that the nested native build uses the same python interpreter as
the main build, in case the python that CMake detects first is not the
python that the user has specified explicitly.

Original fix by Anuj Mittal <anuj.mittal@intel.com>.

DavidSpickett commented:
And this broke the Python formatter, as in, it failed to parse some code at all. Probably one of the infrequently changed projects like Polly has some strange code in it that we've never tried to format. Point is - we know this isn't changing actual Python code so ignore it.

rossburton committed:
[polly] Use python2 interpeter on scripts that are explicitly Python 2

The scripts in polly/utils use Python 2 syntax (notable, print is a
statement not a function), so change the interpeter to be python2 as
they will not run under python3.

rossburton commented:
The PYTHON_EXECUTABLE and the shebang cleanups are entirely unrelated except by theme, I can split them it that would make things easier.

The use-case was "a few scripts are using `/usr/bin/env python` and that broke in builds where the host has no `python`" and instead of playing whack-a-mole and fixing just the files that triggered problems, I blanket-fixed the shebangs.

Apart from a few which the formatter exploded on and are actually python2 scripts, so the current shebang wouldn't work as for ~everyone who has `python`, it's actually python3. This suggests that the polly/utils scripts haven't been ran by anyone for quite some time.

DeinAlptraum commented:
Relevant PEP: https://peps.python.org/pep-0394/

To quote:
"For scripts that are only expected to be run in an activated virtual environment, shebang lines can be written as #!/usr/bin/env python, as this instructs the script to respect the active virtual environment."

So I'm not sure the shebang for venvs should be changed to python3.

Moreover, this may have pretty big implications since the implementation of these commands 8s very inconsistent across platforms and no common rules seem to exist.
Some platforms have `python` that aliases to `python3`, some alias it to `python2`, some only have `python` etc. so I'm not convinced it's a good idea to just change these broadly without considering possible platform specific requirements

mchoo7 reviewed (changes requested):
The FreeBSD part (`copy-parse.py`) has been removed. If you merge main, FreeBSD-related change will be removed from the diff and it won't need my approval.
